File: easel/ast.js
export class Literal {
  constructor(value) {
    this.value = value
  }
}

export class Binary {
  constructor(left, operator, right) {
    this.left
    this.operator = operator
    this.right = right
  }
}

export class Var {
  constructor(name, value) {
    this.name = name
    this.value = value
  }
}

export class Struct {
  constructor(name, members) {
    this.name = name
    this.members = members
  }
}

export class Func {
  constructor(name, params, body) {
    this.name = name
    this.params = params
    this.body = body
  }
}

export class Call {
  constructor(caller, args) {
    this.caller = caller
    this.args = args
  }
}

export class Get {
  constructor(caller, property) {
    this.caller = caller
    this.property = property
  }
}

export class For {
  constructor(id, range, body) {
    this.id = id
    this.range = range
    this.body = body
  }
}

export class While {
  constructor(condition, body) {
    this.condition = condition
    this.body = body
  }
}

export default {
  Literal,
  Binary,
  Var,
  Struct,
  Func,
  Call,
  Get,
  For,
  While
}

File: languages/monopolang/src/ast.rs
use crate::lexer::TokenType;

#[derive(Debug, Copy, Clone)]
pub enum UnaryOperator {
    Negate,
    Not,
}

#[derive(Debug, Copy, Clone)]
pub enum BinaryOperator {
    // Arithmetic
    Add,
    Subtract,
    Multiply,
    Divide,

    // Boolean
    Equal,
    NotEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
}

#[derive(Debug, Copy, Clone)]
pub enum LogicalOperator {
    And,
    Or,
}

#[derive(Debug, Clone)]
pub enum Declaration {
    Statement(Statement),
    Procedure(String, Vec<Statement>), // Name, Code
}

#[derive(Debug, Clone)]
pub enum Statement {
    VariableAssignment(String, Expression), // Name, Value
    Expression(Expression),
    Print(Expression),
    Block(Vec<Statement>),
    If(Expression, Box<Statement>, Option<Box<Statement>>), // Condition, Then, Else
    ProcedureCall(String),                                  // Name
    Gamble(Expression),                                     // Amount to gamble
    Buy(Expression, Expression),                            // Stock, Amount
    Sell(Expression, Expression),                           // Stock, Amount
    Loan(Expression),                                       // Take out loan for amount
    Pay(Expression),                                        // Amount to pay back loan
    While(Expression, Box<Statement>),                      // Condition, Body
    Range(String, Expression, Expression, Expression, Box<Statement>), // Variable name, Start, End, Step, Body
    Work,
}

#[derive(Debug, Clone)]
pub enum Expression {
    Number(f64),
    Boolean(bool),
    Void,
    String(String),
    Variable(String),
    ReadonlyVariable(String), // Used internally for economy variables, like @balance
    StockPrice(String),       // Used to access stock prices from inside the program
    Unary(UnaryOperator, Box<Expression>), // Operator, Operand
    Binary(BinaryOperator, Box<Expression>, Box<Expression>), // Operator, Left, Right
    Logical(LogicalOperator, Box<Expression>, Box<Expression>), // Operator, Left, Right
}

impl UnaryOperator {
    pub fn from_tokentype(kind: TokenType) -> Self {
        match kind {
            TokenType::Bang => Self::Not,
            TokenType::Minus => Self::Negate,
            _ => unreachable!("Invalid unary operator"),
        }
    }
}

impl BinaryOperator {
    pub fn from_tokentype(kind: TokenType) -> Self {
        match kind {
            TokenType::Plus => Self::Add,
            TokenType::Minus => Self::Subtract,
            TokenType::Star => Self::Multiply,
            TokenType::Slash => Self::Divide,
            TokenType::Equal => Self::Equal,
            TokenType::BangEqual => Self::NotEqual,
            TokenType::Greater => Self::Greater,
            TokenType::GreaterEqual => Self::GreaterEqual,
            TokenType::Less => Self::Less,
            TokenType::LessEqual => Self::LessEqual,
            _ => unreachable!("Invalid binary operator"),
        }
    }
}

impl LogicalOperator {
    pub fn from_tokentype(kind: TokenType) -> Self {
        match kind {
            TokenType::And => Self::And,
            TokenType::Or => Self::Or,
            _ => unreachable!("Invalid logical operator"),
        }
    }
}
